File: problems/0853-car-fleet/go/solution.go
package solution

import "sort"

// Car Fleet — https://leetcode.com/problems/car-fleet/
func carFleet(target int, position []int, speed []int) int {
	idx := make([]int, len(position))
	for i := range idx {
		idx[i] = i
	}
	// sort by starting position, descending (closest to target first)
	sort.Slice(idx, func(a, b int) bool {
		return position[idx[a]] > position[idx[b]]
	})

	fleets := 0
	curTime := 0.0
	for _, i := range idx {
		time := float64(target-position[i]) / float64(speed[i])
		if time > curTime {
			fleets++
			curTime = time
		}
	}
	return fleets
}

File: problems/0853-car-fleet/typescript/solution.ts
// Car Fleet — https://leetcode.com/problems/car-fleet/
export function carFleet(target: number, position: number[], speed: number[]): number {
  const cars = position
    .map((pos, i) => ({ pos, speed: speed[i] }))
    .sort((a, b) => b.pos - a.pos); // closest to target first

  let fleets = 0;
  let curTime = 0;
  for (const car of cars) {
    const time = (target - car.pos) / car.speed;
    if (time > curTime) {
      fleets++;
      curTime = time;
    }
  }
  return fleets;
}
